routing/control_tower.go

```diff
@@ -85,10 +141,115 @@ func (p *controlTower) Success(paymentHash lntypes.Hash,
 func (p *controlTower) Fail(paymentHash lntypes.Hash,
 	reason channeldb.FailureReason) error {
 
-	return p.db.Fail(paymentHash, reason)
+	// Lock hash to prevent the race condition described in the Success
+	// method.
+	p.hashMutex.Lock(string(paymentHash[:]))
+	defer p.hashMutex.Unlock(string(paymentHash[:]))
```

Comment:
same, I think it can be locked after the db tx is over. 

Would this also let us use a regular mutex? As the mutex would no longer encompass the critical path (multiple payments recording attempts and results with the DB).

Reply:
See comment on `Success`